datafusion/substrait/src/logical_plan/consumer.rs

```diff
@@ -2037,42 +2438,48 @@ fn from_substrait_type(
                 Ok(DataType::Interval(IntervalUnit::MonthDayNano))
             }
             r#type::Kind::UserDefined(u) => {
-                if let Some(name) = extensions.types.get(&u.type_reference) {
+                if let Ok(data_type) = consumer.consume_user_defined_type(u) {
+                    return Ok(data_type);
+                }
+
+                // TODO: remove the code below once the producer has been updated
```

Comment:
I think the producer doesn't produce any of these anymore, we just didn't remove them yet in case someone is producing plans with an older version and trying to run them with a newer one. Or like, run a plan that has been stored somewhere. But maybe it's time to remove them from the next version for example.

Reply:
I thought about removing it as part of this PR, but figured that would be better handled as a separate follow-up change. I'm already breaking enough API with this as is 😅